contracts/JBGenericBuybackDelegate.sol

```diff
@@ -601,44 +525,37 @@ contract JBGenericBuybackDelegate is
             useReservedRate: true
         });
 
-        emit BuybackDelegate_Swap(_data.projectId, _data.forwardedAmount.value, _amountReceived);
+        emit BuybackDelegate_Swap(_data.projectId, _amountToSwapWith, _pool, _amountReceived, msg.sender);
     }
 
-    /**
-     * @notice Mint the token out, sending back the token in the terminal
-     *
-     * @param  _data the didPayData passed by the terminal
-     * @param  _amount the amount of token out to mint
-     */
-    function _mint(JBDidPayData3_1_1 calldata _data, uint256 _amount) internal {
-        // Mint to the beneficiary with the fc reserve rate
+    /// @notice Add the specified amount of funds back into the project's terminal, and mint the appropriate amount of project tokens.
+    /// @param _data The didPayData passed by the terminal.
+    /// @param _amount The amount to add back to the project's balance.
+    /// @param _weight The relative amount of tokens that should be minted when a project receives funds.
+    function _mint(JBDidPayData3_1_1 calldata _data, uint256 _amount, uint256 _weight) internal {
+        // Keep a reference to the number of tokens being minted.
+        uint256 _tokenCount = mulDiv18(_amount, _weight);
+       
+        // Mint to the beneficiary, making sure the reserved rate gets taken into account.
         CONTROLLER.mintTokensOf({
             projectId: _data.projectId,
-            tokenCount: _amount,
+            tokenCount: _tokenCount,
             beneficiary: _data.beneficiary,
             memo: _data.memo,
             preferClaimedTokens: _data.preferClaimedTokens,
             useReservedRate: true
         });
 
-        // Add the token or eth back to the terminal balance
+        // If the token paid in wasn't ETH, give the terminal permission to pull them back into its balance.
         if (_data.forwardedAmount.token != JBTokens.ETH) {
             IERC20(_data.forwardedAmount.token).approve(msg.sender, _data.forwardedAmount.value);
```

Comment:
I think `_data.forwardedAmount.value` should be `_amount`, the approval amount is for the entire payment amount instead of the `_terminalTokenInThisContract`

Reply:
@xBA5ED Don't we end up with too much allowance then? The idea is just to send the token received back (which are the fwd amount), no?

Reply:
we only transfer _amount though...